### PR TITLE
Fix: Visibility Optimization randomly not working

### PR DESCRIPTION
**Pull request by hiroshihorie**
Should fix issue https://github.com/livekit/client-sdk-flutter/issues/63

Seems to work well but further testing is required.

### CODE REVIEW

**Comment by davidzhao** [comment]
this looks fine, but I'm not really understanding what caused it to fail before. in which scenarios would it fail before?

**Review by davidzhao — APPROVED** [review]

File: lib/src/publication/remote.dart [hunk 1]
```diff
@@ -97,6 +98,10 @@ class RemoteTrackPublication<T extends RemoteTrack>
   // from one of the renderers
   void _onVideoRendererVisibilityUpdateEvent(
       TrackVisibilityUpdatedEvent event) {
+    if (event.track is LocalVideoTrack) {
```
Comment (davidzhao) [inline lib/src/publication/remote.dart:101]: question: why is the RemoteTrackPublication handling events from a LocalVideoTrack?

Comment (hiroshihorie) [reply lib/src/publication/remote.dart:101]: Right! this should never happen so the check is not necessary. I will remove it.

**Comment by hiroshihorie** [comment]
> this looks fine, but I'm not really understanding what caused it to fail before. in which scenarios would it fail before?
It was unstable because Flutter frequently recreates the widget instances and I think the only reliable way to identify a widget is by using `Key`.
But i think i will just use an alternative (simple) method that doesn't rely on widget ids.

**Comment by hiroshihorie** [comment]
re-designed and simplified so doesn't rely on `visibility_detector`, doesn't rely on view ids...

**Comment by hiroshihorie** [comment]
> this looks fine, but I'm not really understanding what caused it to fail before. in which scenarios would it fail before?


It failed before because the id i used to manage VideoViews were unreliable. using objectId doesn't work because Widget gets rebuilt and also State gets reused.

New approach simply scans all visible VideoViews periodically.

**Comment by davidzhao** [comment]
got it. nice work!

**Pull request merged**